### PR TITLE
Avoid overwriting cache with null data

### LINKED ISSUE #1745: Status icons not colored
Seen this a few times, especially on a cold launch. Refreshing fixes it. 

GitHawk Upload by BasThomas

<details>
<summary>Bug Report Dump (Auto-generated)</summary>
<pre>
Version 1.19.0 (1524358860)
Device: iPhone X (iOS 11.3)
TestFlight: true
</pre>
</details>

### LINKED ISSUE #1869: Don’t set the notification # to 0 if loading fails
If you have a certain amount of notifications and then open the app without connection, it will reset the counter to 0.

<details>
<summary>Bug Report Dump (Auto-generated)</summary>
<pre>
Version 1.21.0 (1528075587)
Device: iPhone X (iOS 11.4)
TestFlight: true
</pre>
</details>

### PR DESCRIPTION
**Pull request by rob-nash**
This pull request prevents cache data being overwritten, which closes the following two issues.

Closes #1869 
Closes #1745

### CODE REVIEW

**Comment by rob-nash** [comment]
The function `createNotification` might need to be renamed to something more suitable like `fetchOrCreate`.

I think the `NotificationModelController.swift` file needs some refactoring, but that should be raised as a separate ticket. There are some inline annotations there that suggest moving parsing code somewhere else. I've noticed `self` being captured in places and some repeating code here or there.

File: Classes/Notifications/CreateNotificationTitle.swift [hunk 1]
```diff
@@ -9,7 +9,7 @@
 import Foundation
 import StyledTextKit
 
-func CreateNotification(title: String, width: CGFloat, contentSizeCategory: UIContentSizeCategory) -> StyledTextRenderer {
+func createNotification(title: String, width: CGFloat, contentSizeCategory: UIContentSizeCategory) -> StyledTextRenderer {
```
Comment (rnystrom) [inline Classes/Notifications/CreateNotificationTitle.swift]: Why this change? Not really relevant to the rest of the PR. Could we revert and revisit in another PR if necessary?

File: Classes/Notifications/Notification+NotificationViewModel.swift [hunk 2]
```diff
@@ -8,49 +8,54 @@
 
 import UIKit
 import GitHubAPI
+import FlatCache
 
-func CreateNotificationViewModels(
+func createNotificationViewModels(
```
Comment (rnystrom) [inline Classes/Notifications/Notification+NotificationViewModel.swift]: Same

File: Classes/Notifications/Notification+NotificationViewModel.swift [hunk 3]
```diff
@@ -8,49 +8,54 @@
...
     width: CGFloat,
     contentSizeCategory: UIContentSizeCategory,
     v3notifications: [V3Notification],
+    cache: FlatCache,
     completion: @escaping ([NotificationViewModel]) -> Void
     ) {
-    DispatchQueue.global().async {
```
Comment (rnystrom) [inline Classes/Notifications/Notification+NotificationViewModel.swift]: Why are we removing these? Wont this execute on the main thread? The models these methods generate include warming text size caches, expensive!

File: Classes/Notifications/CreateNotificationTitle.swift (hunk 1 above)
Comment (rob-nash) [reply Classes/Notifications/CreateNotificationTitle.swift]: Ok

File: Classes/Notifications/Notification+NotificationViewModel.swift (hunk 2 above)
Comment (rob-nash) [reply Classes/Notifications/Notification+NotificationViewModel.swift]: Yep

File: Classes/Notifications/Notification+NotificationViewModel.swift (hunk 3 above)
Comment (rob-nash) [reply Classes/Notifications/Notification+NotificationViewModel.swift]: I was unsure about this one. The cache need the main queue and I couldn't see an obvious reason why this Dispatch was occurring. 

Will add it back in. 👍

File: Classes/Notifications/NotificationModelController.swift [hunk 4]
```diff
@@ -58,12 +58,14 @@ final class NotificationModelController {
...
                 case .success(let response):
+
                     badge.updateLocalNotificationCache(notifications: response.data, showAlert: false)
 
                     CreateNotificationViewModels(
                         width: width,
                         contentSizeCategory: contentSizeCategory,
-                        v3notifications: response.data
+                        v3notifications: response.data,
+                        cache: self.githubClient.cache
```
Comment (rnystrom) [inline Classes/Notifications/NotificationModelController.swift]: `self` should be `weak`

File: Classes/Notifications/NotificationModelController.swift [hunk 5]
```diff
@@ -80,7 +82,8 @@ final class NotificationModelController {
...
```
Comment (rnystrom) [inline Classes/Notifications/NotificationModelController.swift]: same

File: Classes/Notifications/NotificationModelController.swift (hunk 4 above)
Comment (rob-nash) [reply Classes/Notifications/NotificationModelController.swift]: Damn, missed that.

File: Classes/Notifications/Notification+NotificationViewModel.swift [hunk 6]
```diff
@@ -8,49 +8,73 @@
...
-                let identifier = notification.subject.identifier
-                else { continue }
-
+        
+        for v3notification in uncachedV3Notifications {
+            
+            guard
+                let type = NotificationType(rawValue: v3notification.subject.type.rawValue),
+                let identifier = v3notification.subject.identifier else {
+                    return
```
[the remark this answers is not in the document]
Comment (rob-nash) [reply Classes/Notifications/Notification+NotificationViewModel.swift]: You're right

File: Classes/Notifications/Notification+NotificationViewModel.swift [hunk 7]
```diff
@@ -8,49 +8,73 @@
...
-            ))
+            )
+            
+            newModels.append(model)
         }
-
+        
         DispatchQueue.main.async {
+            models.append(contentsOf: cachedModels)
+            models.append(contentsOf: newModels)
```
[the remark this answers is not in the document]
Comment (rob-nash) [reply Classes/Notifications/Notification+NotificationViewModel.swift]: Yes

**Comment by rob-nash** [comment]
You're right. I keep making mistakes. I'll close it.
